File: 1568.cpp
/**
1568. Minimum Number of Days to Disconnect Island
https://leetcode.com/problems/minimum-number-of-days-to-disconnect-island/
TLE
**/
class Solution {
private:
    bool isOneIsland(vector<vector<int>> &grid) {
        int n = grid.size(), m = grid[0].size();
        int count = 0;
        // vector<vector<bool>> vis = (n, vector<bool>(m, false));
        set<pair<int,int>> vis;
        for (int i = 0; i < n; ++i)
            for (int j = 0; j < m; ++j)
                if (grid[i][j] == 1 && !vis.count(make_pair(i,j))) {
                    bfs(grid, i, j, vis);
                    ++count;
                }
        return count == 1;
    }
    
    void bfs(vector<vector<int>> &grid, int x, int y, set<pair<int,int>>& v) {
        int n = grid.size(), m = grid[0].size();
        int delta_x[4] = {0, 0, -1, 1};
        int delta_y[4] = {-1, 1, 0, 0};
        deque<pair<int,int>> q;
        q.push_back(make_pair(x,y));
        v.insert(make_pair(x,y));
        while (!q.empty()) {
            pair<int,int> f = q.front(); q.pop_front();
            for (int i = 0; i < 4; ++i) {
                int nx = f.first + delta_x[i], ny = f.second + delta_y[i];
                if (!isValid(nx, ny, grid, v)) continue;
                v.insert(make_pair(nx,ny));
                q.push_back(make_pair(nx,ny));
            }
        }
        return;
    }
    
    bool isValid(int x, int y, vector<vector<int>> &grid, set<pair<int,int>> v) {
        if (x < 0 || x >= grid.size()) return false;
        if (y < 0 || y >= grid[0].size()) return false;
        if (v.count(make_pair(x,y))) return false;
        if (grid[x][y] == 0) return false;
        return true;
    }
public:
    int minDays(vector<vector<int>>& grid) {
        if (!isOneIsland(grid)) return 0;
        for (int i = 0; i < grid.size(); ++i) {
            for (int j = 0; j < grid[0].size(); ++j) {
                if (grid[i][j] == 1) {
                    grid[i][j] = 0;
                    if (!isOneIsland(grid)) return 1;
                    grid[i][j] = 1;
                }
            }
        }
        return 2;
    }
};

File: 1568.py
"""
1568. Minimum Number of Days to Disconnect Island
https://leetcode.com/problems/minimum-number-of-days-to-disconnect-island/
"""
class Solution:
    def minDays(self, grid: List[List[int]]) -> int:
        land = self.count_land(grid)
        if not self.isOneIsland(grid, land):
            return 0
        n, m = len(grid), len(grid[0])
        for i in range(n):
            for j in range(m):
                if grid[i][j] == 1:
                    grid[i][j] = 0
                    if not self.isOneIsland(grid, land - 1):
                        return 1
                    grid[i][j] = 1
        return 2
    
    def count_land(self, grid):
        n, m = len(grid), len(grid[0])
        count = 0
        for i in range(n):
            for j in range(m):
                if grid[i][j] == 1:
                    count += 1
        return count
    
    def isOneIsland(self, grid, land):
        n, m = len(grid), len(grid[0])
        count = 0
        for i in range(n):
            for j in range(m):
                if grid[i][j] == 1:
                    count = self.bfs(grid, i, j)
                    return count == land
        return False
    
    def bfs(self, grid, x, y):
        DIRECTIONS = [
            (0, 1),
            (0, -1),
            (-1, 0),
            (1, 0)
        ]
        n, m = len(grid), len(grid[0])
        q = deque([(x,y)])
        v = set([(x,y)])
        count = 0
        while q:
            x, y = q.popleft()
            count += 1
            for delta_x, delta_y in DIRECTIONS:
                nx, ny = x + delta_x, y + delta_y
                if not self.is_valid(nx, ny, grid, v):
                    continue
                q.append((nx,ny))
                v.add((nx,ny))
        return count
    
    def is_valid(self, x, y, grid, v):
        n, m = len(grid), len(grid[0])
        if (x < 0 or x >= n or y < 0 or y >= m):
            return False
        if (x, y) in v:
            return False
        if grid[x][y] == 0:
            return False
        return True